### PR TITLE
Linux: Reboot Reason Tracking documentation

### PR DESCRIPTION
**Pull request by martijnthe**
Adds documentation for the revamped reboot reason tracking in the [Linux SDK 1.1.0](https://github.com/memfault/memfault-linux-sdk/tree/1.1.0-kirkstone).

FireShot Capture 002 - Linux Reboot Reason Tracking - Memfault Docs - localhost

### CODE REVIEW

File: docs/mcu/reboot-reason-tracking.mdx [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ---
 id: reboot-reason-tracking
-title: Reboot Reason Tracking
+title: MCU Reboot Reason Tracking
```
Comment (martijnthe) [inline docs/mcu/reboot-reason-tracking.mdx:3]: I renamed this to make the search results a bit more useful (the search doesn't show the section/category a page is part of).

File: docs/platform/reference-reboot-reason-ids.mdx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+information about reboot reason tracking, check out one of these pages:
+
+- [Reboot Reason Tracking (MCU Guide)][docs-mcu-reboot-reason-tracking]
+- [Reboot Reason Tracking (Android Guide)][docs-android-reboot-reason-tracking]
+- [Reboot Reason Tracking (Linux Guide)][docs-linux-reboot-reason-tracking]
+
+[docs-mcu-reboot-reason-tracking]: /docs/mcu/reboot-reason-tracking
+[docs-android-reboot-reason-tracking]: /docs/android/android-reboot-events
+[docs-linux-reboot-reason-tracking]: /docs/linux/reboot-reason-tracking
+
```
Comment (martijnthe) [inline docs/platform/reference-reboot-reason-ids.mdx:17]: I pulled these names from the frontend code, aka what customers see in the web app.

I noticed that some of the names deviate from the [MCU SDK enum types](https://github.com/memfault/memfault-firmware-sdk/blob/master/components/include/memfault/core/reboot_reason_types.h).
Maybe worth aligning them?

cc @noahp @ejohnso49 @chrisc11

**Comment by fnune** [comment]
For @sarfata: you can see a rendered version of this PR here https://deploy-preview-338--memfault-docs.netlify.app/docs/linux/reboot-reason-tracking

You can find it in other PRs by clicking here:
image

File: docs/linux/reboot-reason-tracking.mdx [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+set this up for your device. A key function of `memfaultd` is to detect,
+classify and upload reboot events to the Memfault platform. It does this via its
+`plugin_reboot`.
+
+`plugin_reboot` is enabled by default. Read more on [how to configure which
+plugins `memfaultd` builds with][docs-linux-control-plugins].
+
+### Linux kernel `pstore` / `ramoops` configuration
+
+The detection of kernel panics depends on the so called `ramoops` subsystem and
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:49]:
```suggestion
The detection of kernel panics as a reboot reason depends on the so called `ramoops` subsystem and
```

Nit: may be confusing to read "detection of kernel panics" since we don't provide logs/crash reports for those, but just figure out the reboot reason.

File: docs/linux/reboot-reason-tracking.mdx [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+`pstore` filesystem. From the [Linux kernel admin guide][ramoops-guide]:
+
+> Ramoops is an oops/panic logger that writes its logs to RAM before the system
+> crashes. It works by logging oopses and panics in a circular buffer. Ramoops
+> needs a system with persistent RAM so that the content of that area can
+> survive after a restart.
+
+The `pstore` is a RAM-backed filesystem that persists across reboots.
+
+The easiest way to enable the `pstore` in your kernel is via the
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:59]: Nit: add "if using Yocto" somewhere? In some other docs we have tabs. If we don't have instructions that are build-system-agnostic, then let's indicate that this is about Yocto.

File: docs/linux/reboot-reason-tracking.mdx [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+There are only a few reboot reasons that can be determined with confidence for
+all embedded Linux devices. The reboot reasons that memfaultd can determine are
+these reboots:
+
+| Reason          | Description                                  | Information Source          |
+| --------------- | -------------------------------------------- | --------------------------- |
+| Kernel Panic    | The Linux kernel crashed                     | `pstore` / `ramoops`        |
+| Firmware Update | The device rebooted because of an OTA update | U-Boot environment variable |
+| User Reset      | "Graceful" shutdown of the system            | `systemd` state             |
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:164]: Add a reminder that Firmware Update is only available if you've integrated with OTA and must otherwise be implemented manually?

File: docs/linux/reboot-reason-tracking.mdx [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+# Must match reboot_plugin.uboot_fw_env_file in memfaultd.conf:
+CONFIG_UBOOT_FWENV="/etc/fw_env.config"
+CONFIG_BOOTLOADER_DEFAULT_UBOOT=y
+CONFIG_UPDATE_STATE_CHOICE_BOOTLOADER=y
+CONFIG_UPDATE_STATE_BOOTLOADER="ustate"
+```
+
+## Reboot Reason Classification
+
+### Built-in
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:154]:
```suggestion
### Built-in reboot reasons
```

May sound redundant but will make the anchor and the TOC more readable.

File: docs/linux/reboot-reason-tracking.mdx [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+all embedded Linux devices. The reboot reasons that memfaultd can determine are
+these reboots:
+
+| Reason          | Description                                  | Information Source          |
+| --------------- | -------------------------------------------- | --------------------------- |
+| Kernel Panic    | The Linux kernel crashed                     | `pstore` / `ramoops`        |
+| Firmware Update | The device rebooted because of an OTA update | U-Boot environment variable |
+| User Reset      | "Graceful" shutdown of the system            | `systemd` state             |
+
+### Device-specific
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:166]:
```suggestion
### Device-specific reboot reasons
```

Why not call them custom reboot reasons instead?

Also, ditto about this:

> May sound redundant [to add " reboot reasons"] but will make the anchor and the TOC more readable.

File: docs/linux/reboot-reason-tracking.mdx [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+`reboot_plugin.last_reboot_reason_file` key in `memfaultd.conf`. Take a look at
+the [`/etc/memfaultd.conf` reference][docs-configuration] for more information.
+
+During the startup of the system, `memfaultd` will attempt to read this file. If
+the `last_reboot_reason_file` file exists, it is expected to contain the decimal
+[reboot reason ID][docs-reboot-reason-ids] that corresponds to the _last_ reboot
+that occurred. If the file does not exist, `memfaultd` will interpret this as
+"no device specific reason known".
+
+Generally speaking, there are 2 points during a boot cycle where your software
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:196]:
```suggestion
There are 2 points during a boot cycle where your software
```

Nit: "generally speaking" sounds friendly but this is kind of a reference-style doc so I would make sure it sounds assertive.

File: docs/linux/reboot-reason-tracking.mdx [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+   User Reset)
+
+In other words, a Kernel Panic always trumps other sources. After that, any
+reason found in `last_reboot_reason_file` (even if it is the "Unspecified"
+reason) will trump the built-in reboot reason detection.
+
+If none of the sources determined a reason for a reboot, "Unspecified" will be
+used as the reason.
+
+### Enqueuing & Upload
```
Comment (fnune) [inline docs/linux/reboot-reason-tracking.mdx:232]: OT: English is so weird... why isn't it "enqueu**e**ing"? :smile:

**Review by fnune — APPROVED** [review]
This is awesome! Love the graphics.

I mostly have nitpicks. Here are some more nitpicks about the diagram:
image

**Review by HBehrens — APPROVED** [review]
Looks good generally. A couple of inline suggestions and comments.

Also, could you update the Linux Introduce page as part of the PR, please? The section "Reboot Reason Tracking" could benefit from some more verbiage and a link to this new page.

I haven't followed the instructions myself and only mentally tried to put myself in the readers shoes. Does it make sense for someone, e.g. @sarfata to test this out for real?

Will approve already to unblock the next steps – trusting you will address the feedback where reasonable. Great work, @martijnthe , thank you!

File: docs/linux/reboot-reason-tracking.mdx [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+There are many types of reboots for which the detection can only be implemented
+in device-specific ways.
+
+For example, an SoC may have special hardware to detect power brownouts. There
+is no "standard" way to read brownout detection state in Linux. Including
+implementations in the SDK for every possible device-specific reboot reason is
+unfeasible.
+
+Another example: a device's software may contain logic to decide that the device
```
Comment (sarfata) [inline docs/linux/reboot-reason-tracking.mdx:176]: As a reader, new to all of this, I understand how the system works but I don't see a reboot reason for low battery and it's not clear to me whether I should just make my own integer value and use it (what happens in the UI if I pick a value not in the file?) or do something else.

I think the low battery example is great and we should continue to develop that example in the steps below. 

I am not sure how "low battery shutdown" is usually handled in a linux device but we could imagine including a modified `low-battery-detection.sh` file which in essence would do: 
```
if [ -f /media/shutdown-for-low-battery ]; then
  echo "4242" > /media/last_reboot_reason
fi
```
And we could provide the `low-battery-detection.service` which would include the `Before=memfaultd.service`.

File: docs/linux/reboot-reason-tracking.mdx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+### Reboot Reason Precedence
+
+As described in the previous sections, there are multiple sources of information
+where reboot reasons may come from. In case there are multiple reasons found,
+this order of precedence is used:
+
+1. `pstore`/`ramoops` (Kernel Panic)
+2. `last_reboot_reason_file`
+3. non-`pstore`/`ramoops`, built-in reboot reason detection (Firmware Update or
```
Comment (sarfata) [inline docs/linux/reboot-reason-tracking.mdx:222]: `built-in` is confusing here because we have a paragraph named `built-in` above that includes `pstore/ramoops`. I understand you are saying "non-pstore builtin" but it's a bit hard to follow.

Maybe 3 is just "Firmware update (swupdate) or User reset (systemd)"
```suggestion
3. Firmware update (detected via `swupdate`) or User reset (via `systemd` state)
```

File: docs/linux/reboot-reason-tracking.mdx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+   User Reset)
+
+In other words, a Kernel Panic always trumps other sources. After that, any
```
Comment (sarfata) [inline docs/linux/reboot-reason-tracking.mdx:225]: I would remove this paragraph. The list is above is plenty clear IMHO. If you do want to keep it, rephrase to avoid the use of "built-in reboot reason".

File: docs/linux/reboot-reason-tracking.mdx [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+Once `memfaultd` determines the reboot reason, it will enqueue a reboot event
+for upload to the Memfault cloud. Note that uploading does not happen
+immediately. Instead, the queue of events is uploaded periodically. See the
+[`memfaultd.conf` configuration key
+`refresh_interval_seconds`][docs-configuration] for more information.
+
+:::note Rate Limiting
+
+Ingestion of Reboots Events may be
+[rate-limited](/docs/platform/rate-limiting#reboot-events).
```
Comment (sarfata) [inline docs/linux/reboot-reason-tracking.mdx:243]: The documentation we are pointing to here lists Android and MCU but does not mention Linux.

File: docs/platform/reference-reboot-reason-ids.mdx [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+| ----------- | ------------ | -------- |
+| Usage Fault | 37632        | 0x9300   |
+| Undefinstr  | 37633        | 0x9301   |
+| Invstate    | 37634        | 0x9302   |
+| Invpc       | 37635        | 0x9303   |
+| Nocp        | 37636        | 0x9304   |
+| Unaligned   | 37637        | 0x9305   |
+| Divbyzero   | 37638        | 0x9306   |
+
+### Misc.
```
[the remark this answers is not in the document]
Comment (sarfata) [reply docs/platform/reference-reboot-reason-ids.mdx:88]: +1 yes. This is one of my general comment. What code should i use for low battery reboot?

File: docs/linux/reboot-reason-tracking.mdx (hunk 5 above)
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:165]: Added this in the table:

```
Note: requires [integrating OTA][docs-ota].
```

And also added this to the "SWUpdate configuration" section on this page:

```

:::note

In case you decide not to [use SWUpdate for OTA][docs-ota], you can use
[the `last_reboot_reason_file` API (described in the next section)](#device-specific)
to associate the "Firmware Update" reason to a reboot after your OTA agent has
installed an update.

:::
```

File: docs/linux/reboot-reason-tracking.mdx (hunk 7 above)
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:167]: To me, "custom" sounds like you can customize the reason itself.

**Comment by fnune** [comment]
@martijnthe I'm noticing this a bit late, but the docs never mention what one should write to the `last_reboot_reason_file`? Ints, strings? Do we have a spec?

File: docs/platform/reference-reboot-reason-ids.mdx [hunk 15]
```diff
@@ -0,0 +1,93 @@
+---
+id: reference-reboot-reason-ids
+title: "Reference: Reboot Reason IDs"
+sidebar_label: "Reboot Reason IDs"
+---
+
+This page documents identifiers for different reboot reasons. For more
+information about reboot reason tracking, check out one of these pages:
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/platform/reference-reboot-reason-ids.mdx:8]: Android doesn't have an API for this at all.

File: docs/linux/reboot-reason-tracking.mdx [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+
+The location of this file can be configured using the
+`reboot_plugin.last_reboot_reason_file` key in `memfaultd.conf`. Take a look at
+the [`/etc/memfaultd.conf` reference][docs-configuration] for more information.
+
+During the startup of the system, `memfaultd` will attempt to read this file. If
+the `last_reboot_reason_file` file exists, it is expected to contain the decimal
+[reboot reason ID][docs-reboot-reason-ids] that corresponds to the _last_ reboot
+that occurred. If the file does not exist, `memfaultd` will interpret this as
+"no device specific reason known".
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:195]: memfaultd logs what source was used (and which ones were ignored too).
I'll mention that.

[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:195]: Adding

```

:::note

To aid debugging, `memfaultd` will log the source of the reboot reason that was
used as well as any sources found that provided a reason, but got discarded
because another source took precedence over it:

```
$ journalctl -u memfaultd
Nov 15 09:15:36 qemuarm64 systemd[1]: Starting memfaultd daemon...
Nov 15 09:15:36 qemuarm64 memfaultd[184]: reboot:: Using reboot reason ButtonReset (0x0006) from custom source for boot_id 847c0502-9f47-43d0-b58d-7029b30f9cd0
Nov 15 09:15:36 qemuarm64 memfaultd[184]: reboot:: Discarded reboot reason UserReset (0x0002) from internal source for boot_id 847c0502-9f47-43d0-b58d-7029b30f9cd0
...
```

:::
```

File: docs/linux/reboot-reason-tracking.mdx [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+  during shutdown.
+
+### Reboot Reason Precedence
+
+As described in the previous sections, there are multiple sources of information
+where reboot reasons may come from. In case there are multiple reasons found,
+this order of precedence is used:
+
+1. `pstore`/`ramoops` (Kernel Panic)
+2. `last_reboot_reason_file`
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:222]: Added " After reading the file, `memfaultd` will
immediately unlink the file." to the section about `last_reboot_reason_file`.

File: docs/linux/reboot-reason-tracking.mdx [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+[systemd-pstore.service]:
+  https://www.freedesktop.org/software/systemd/man/systemd-pstore.service.html
+[systemd-bbappend]:
+  https://github.com/memfault/memfault-linux-sdk/blob/-/meta-memfault/recipes-core/systemd/systemd_%25.bbappend
+
+### SWUpdate configuration
+
+The Memfault SDK relies on the SWUpdate OTA agent to download and install OTA
+updates. The reboot plugin can detect when a reboot is due to an OTA update
+being installed. For this to work, SWUpdate need to be configured correctly.
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:132]: Ok, explained in more words how it works and what the dependency is.

File: docs/linux/reboot-reason-tracking.mdx [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+
+Another example: a device's software may contain logic to decide that the device
+needs to shut down or reset, e.g. because the battery has dropped below a
+certain point, or because the user pressed a button.
+
+To track reboot reasons which the Memfault SDK cannot possibly know how to
+detect, the SDK provides a way to extend the reboot reason classification, via
+the `last_reboot_reason_file`.
+
+#### The `last_reboot_reason_file` API
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:185]: > Does it make sense to show an example file? ... This would also be a good place to link to the list of known reboot reason ids.

Yes, good point. Will add.

> Does it incorporate boot_id? 

No.

> I don't see that mentioned anywhere on this page but in the diagram. 

Hmm yeah good point. It's kind of an implementation detail. I've removed boot_id from the diagram and rephrased it to: "Has reboot been tracked already? Y/N"?

File: docs/linux/reboot-reason-tracking.mdx (hunk 10 above)
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:177]: I've added a small code example now. Does that make things more clear?

File: docs/linux/reboot-reason-tracking.mdx [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+may write to this file:
+
+- **During system start-up, _before_ `memfaultd` starts:** this is the time to
+  check external reason sources, for example, the brownout detection registers
+  of the SoC. Make sure to clear the original source, after having written the
+  reboot reason to the `last_reboot_reason_file`. You also need to make sure
+  that this code runs _before_ `memfaultd` starts. The recommended way to do is
+  to add [`Before=memfaultd.service` to your `systemd` unit
+  configuration][systemd-before].
+- **During system shut-down:** for controlled, "expected" shutdowns and resets
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:207]: No, we're not logging that at the moment.

File: docs/linux/reboot-reason-tracking.mdx (hunk 13 above)
Comment (martijnthe) [reply docs/linux/reboot-reason-tracking.mdx:244]: Good catch. Fixed.

File: docs/platform/reference-reboot-reason-ids.mdx [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+- [Reboot Reason Tracking (MCU Guide)][docs-mcu-reboot-reason-tracking]
+- [Reboot Reason Tracking (Android Guide)][docs-android-reboot-reason-tracking]
+- [Reboot Reason Tracking (Linux Guide)][docs-linux-reboot-reason-tracking]
+
+[docs-mcu-reboot-reason-tracking]: /docs/mcu/reboot-reason-tracking
+[docs-android-reboot-reason-tracking]: /docs/android/android-reboot-events
+[docs-linux-reboot-reason-tracking]: /docs/linux/reboot-reason-tracking
+
+## Normal Reboot Reasons
```
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/platform/reference-reboot-reason-ids.mdx:18]: I copied the values over from the .h file and the names from ResetReason.ts.

File: docs/platform/reference-reboot-reason-ids.mdx (hunk 14 above)
[the remark this answers is not in the document]
Comment (martijnthe) [reply docs/platform/reference-reboot-reason-ids.mdx:88]: > This is probably out of scope here but should the document discuss reserved ranges or how (currently) unknown values will be interpreted?

Hmm.. I don't know! Leaving it for another day.

> What code should i use for low battery reboot?

Low Power == 4

**Comment by martijnthe** [comment]
> @martijnthe I'm noticing this a bit late, but the docs never mention what one should write to the `last_reboot_reason_file`? Ints, strings? Do we have a spec?

It does mention it?

> If the `last_reboot_reason_file` file exists, it is expected to contain the decimal
[reboot reason ID][docs-reboot-reason-ids] that corresponds to the _last_ reboot
that occurred.
